src/ctap/storage/key.rs

```diff
@@ -73,6 +73,35 @@ make_partition! {
     // - When adding a (non-persistent) key below this message, make sure its value is bigger or
     //   equal than NUM_PERSISTENT_KEYS.
 
+    // Start of key arrays for multi-PIN feature: these fields are separated for each slots, so
+    // a unique key is needed for each slot. However, we reuse the existing fields and rename them
+    // to `FIRST_{KEY_NAME}` so the upgrade is backward compatible.
+    // Depending on `Customization::slot_count()`, only a prefix of those keys is used.
+
+    /// Whether the PIN needs to be changed for each slot, except the first.
+    ///
+    /// If this entry exists and is empty, the PIN needs to be changed.
+    FORCE_PIN_CHANGE = 972..978;
+
+    /// The number of PIN retries for each slot, except the first.
+    ///
+    /// Depending on `Customization::slot_count()`, only a prefix of those keys is used.
+    /// If the entry is absent, the number of PIN retries is `Customization::max_pin_retries()`.
+    PIN_RETRIES = 979..985;
```

Comment:
These ranges are inclusive for the start and exclusive for the end, right? So they are length 6? Can we:

- Add a test that checks that all of these ranges have the same size.
- Make them consecutive?

@ia0 correct? Also, since this restricts the maximum number of slot IDs, we should be confident that 6 is a good number? (Or 7, since we could make the ranges 1 bigger right now.)

Alternative for all slow-changing values like `FORCE_PIN_CHANGE`: Put them behind a single key, and use an array. And make the others a bit bigger instead.

Reply:
Yes, the ranges use the same notation as Rust.

I agree we should not be afraid to allocate more than we currently need to anticipate possible increase in the future. You can do the same as for credentials and have 2 consecutive ranges: `XXX: A..B` and `_RESERVED_XXX: B..C`. We could take `C - A = 16`.

Reply:
Also +1 for using arrays for slow-changing values (in particular when they are small).

Reply:
I think it's my mistake here when arranging the numbers. I meant to make length 8 and consecutive.

Reply:
I largely revised the storage structure to directly concat fields that are just u8 (2 of them) and a u32 field (global signature counter). Dealing with u32 field is more tricky and as @ia0 said maybe a library to more generically handle those concatenated fields are better.
I kept using separated keys for pin properties because I'm not comfortable with "directly" concatenating it without any separator/structure.

Welcome for suggestions though, I'm not really satisfied with the current state of the multi-slot storage and I hope we can come up with something both simple enough and generic, future-proof enough.

Reply:
Seems to be resolved.